logger/analytics/analyticslogger.go

```diff
@@ -89,22 +115,66 @@ func (al *Logger) Write(bs []byte) (int, error) {
 		return 0, err
 	}
 	bs = append(bs, '\n')
-	r := retrier.New(retrier.ExponentialBackoff(5, 100*time.Millisecond), RequestErrorClassifier{})
-	if err := r.Run(func() error {
-		_, err := al.fhAPI.PutRecord(&firehose.PutRecordInput{
-			DeliveryStreamName: aws.String(al.fhStream),
-			Record:             &firehose.Record{Data: bs},
-		})
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return 0, fmt.Errorf("PutRecords: %v", err)
+	al.batchBytes += len(bs)
+	al.batch = append(al.batch, &firehose.Record{Data: bs})
+	if len(al.batch) == al.maxBatchRecords || al.batchBytes > int(0.9*float64(al.maxBatchBytes)) {
+		return len(bs), al.sendBatch()
 	}
 	return len(bs), nil
 }
 
+// Close flushes all logs to Firehose.
+func (al *Logger) Close() error {
+	al.mu.Lock()
+	defer al.mu.Unlock()
+	if len(al.batch) > 0 {
+		return al.sendBatch()
+	}
+	return nil
+}
+
+func (al *Logger) sendBatch() error {
+	defer func() {
+		// reset the batch state no matter what, since
+		// errors will only get worse if we keep adding to
+		// the batch (i.e. we'll exceed record or byte limits)
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+	}()
+
+	// call PutRecordBatch until all records in the batch have been sent successfully
+	for batch := al.batch; ; {
```

Comment:
Do we really want to loop forever until we send everything? Definitely seems bad if we hold the lock the whole time, but even if we don't maybe we want to have some point where we give up.